Improve validation errors doc (#5733)
pydantic/pydantic: docs/usage/validation_errors.md

## `arguments_type`

This error is raised when arguments passed to a function are not

> a tuple, list or a dictionary.

Reviewer: ```suggestion
a tuple, list or dictionary.
```

## `assertion_error`

> This error is raised when there is a failing assertion during the validation.

Reviewer: ```suggestion
This error is raised when a failing assert statement is encountered during validation.
```

```py test="skip" lint="skip" upgrade="skip"
from pydantic import BaseModel, field_validator

class Model(BaseModel):
    x: int

    @field_validator('x')
    @classmethod
    def force_x_positive(cls, v):
        assert v > 0
        return v

Model(x=-1)
```

## `bool`

## `bool_parsing`

> This error is raised when the value is an invalid boolean string.

Reviewer: ```suggestion
This error is raised when the value is a string that is not valid for coercion to a boolean.
```

Author: only string I think

```py test="skip" lint="skip" upgrade="skip"
from pydantic import BaseModel

class Model(BaseModel):
    x: bool

> print(Model(x='test'))

Reviewer: ```suggestion
print(Model(x='true'))  # OK
print(Model(x='test'))
```
I think it's maybe worth making clear via this kind of simple example here that some strings _are_ acceptable

from pydantic import BaseModel

class Model(BaseModel):
    x: bytes

Model(x=123)
```

## `callable_type`

This error is raised when the value is not a `callable`.

```py test="skip" lint="skip" upgrade="skip"
from typing import Any, Callable

from pydantic import BaseModel, ImportString

class Model(BaseModel):
    x: ImportString[Callable[[Any], Any]]

Model(x='os.path')
```

Valid example:

```py test="skip" lint="skip" upgrade="skip"
from typing import Any, Callable

from pydantic import BaseModel, ImportString

class Model(BaseModel):
    x: ImportString[Callable[[Any], Any]]

Model(x='math.cos')
```

## `dataclass_type`

This error is raised when the value is not valid for a dataclass field.

```py test="skip" lint="skip" upgrade="skip"
from pydantic import dataclasses

@dataclasses.dataclass
class Nested:
    x: int

@dataclasses.dataclass
class Model:
    y: Nested

Model(y=1)
```

Valid example:

```py test="skip" lint="skip" upgrade="skip"
from pydantic import dataclasses

@dataclasses.dataclass
class Nested:
    x: int

@dataclasses.dataclass
class Model:
    y: Nested

Model(y=Nested(x=1))
```

## `date_from_datetime_inexact`

> This error is raised when the datetime value provided for a date field is not an exact date.

Reviewer: ```suggestion
This error is raised when the datetime value provided for a date field has a nonzero time component.
```

I'm not sure if my above comment is exactly right, but I think to a newcomer it may not necessarily be clear what is meant by "an exact date". In particular, if someone is opening this link, they were probably hoping for more clarity than that phrase provides, I think we can explain a bit more clearly here. (Something like "For a timestamp to parse into a field of type `date`, the time components must be all zero, and show an example that doesn't fail below.)

```py test="skip" lint="skip" upgrade="skip"
from datetime import date, datetime

from pydantic import BaseModel

class Model(BaseModel):
    x: date

Model(x=datetime(2023, 1, 1, 12))
```

## `date_from_datetime_parsing`

> This error is raised when the datetime value provided for a date field is invalid.

Author: Yes, I think only for `string`.
Changed it.

```py test="skip" lint="skip" upgrade="skip"
from datetime import date

from pydantic import BaseModel

class Model(BaseModel):
    x: date

Model(x='XX1494012000')
```

## `date_future`

This error is raised when the value provided for a `FutureDate` field is not in the future.

```py test="skip" lint="skip" upgrade="skip"
from datetime import date

from pydantic import BaseModel, FutureDate

class Model(BaseModel):
    x: FutureDate

Model(x=date(2000, 1, 1))
```

## `date_parsing`

This error is raised when the value for is not a valid json value for a date field.

```py test="skip" lint="skip" upgrade="skip"
import json
from datetime import date

from pydantic import BaseModel, Field

class Model(BaseModel):
    x: date = Field(strict=True)

Model.model_validate_json(json.dumps({'x': '1'}))
```

## `date_past`

This error is raised when the value provided for a `PastDate` field is not in the past.

```py test="skip" lint="skip" upgrade="skip"
from datetime import date, timedelta

from pydantic import BaseModel, PastDate

class Model(BaseModel):
    x: PastDate

Model(x=date.today() + timedelta(1))
```

## `date_type`

> This error is raised when the value type is not date for a strict date field.

Reviewer: ```suggestion
This error is raised when the value type is not of type `date` for a strict date field.
```
